File: leetcode/#144.py
# question --> https://leetcode.com/problems/binary-tree-preorder-traversal/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        stack = []; 
        stack.append(root); 
        ans = []; 
        while stack: 
            node = stack.pop(); 
            if node:
                ans.append(node.val); 
                if node.right: 
                    stack.append(node.right)
                if node.left: 
                    stack.append(node.left)
        return ans
        
        # TC: O(N); SC: O(N)
        
# end 
        

File: leetcode/#144.js
// question --> https://leetcode.com/problems/binary-tree-preorder-traversal/

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var preorderTraversal = function(root) {
    if (!root) {
        return [];
    }
    let stack = [], ans = [];
    stack.push(root); 
    while (stack.length) {
        let node = stack.pop(); 
        ans.push(node.val); 
        if (node.right) {
            stack.push(node.right); 
        } 
        if (node.left) {
            stack.push(node.left); 
        }
    }
    return ans; 
};

// TC: O(N); SC: O(N)

// end 